### ISSUE TITLE
Rule suggestion: flag non-null assertion applied to an optional chain

### ISSUE THREAD

**Issue by bradzacher** [typescript-eslint/typescript-eslint#1398]
The optional chain operator is built such that it will return `undefined` if there is a nullish member.

Ending an optional chain expression with a non-null assertion is _blatantly_ incorrect, and leads to type safety holes in your code, which will likely lead to runtime errors.

It is much better to either hoist the non-null assertion up the chain, so the thing you "know" is nullish is asserted on, instead of the entire expression.

```ts
function foo(x?: { a?: () => {} }) {
    return x?.a!;
}
function bar(x?: { a?: () => {} }) {
    return x?.a()!;
}
```
